### PR TITLE
plan: decorrelate for aggregation.

### PR DESCRIPTION
**Pull request by hanfei1991**
The rule of decorrelation is
image
This PR only implement Identity(9).

### CODE REVIEW

**Comment by hanfei1991** [comment]
@shenli @coocood @zimulala PTAL

**Comment by shenli** [comment]
@hanfei1991 Could explain the formulations for us?

**Comment by hanfei1991** [comment]
@shenli http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.563.8492&rep=rep1&type=pdf

These cannot be explained by several words.

**Comment by coocood** [comment]
There are many branches in `canPullUp` and `canPullUpAgg`, we need more test cases cover them.

**Comment by coocood** [comment]
LGTM

File: plan/decorrelate.go [hunk 1]
```diff
@@ -102,8 +133,29 @@ func (s *decorrelateSolver) optimize(p LogicalPlan, _ context.Context, _ *idAllo
 				return proj, nil
 			}
 			return s.optimize(p, nil, nil)
+		} else if agg, ok := innerPlan.(*Aggregation); ok {
+			if apply.canPullUpAgg() && agg.canPullUp() {
+				innerPlan = agg.children[0].(LogicalPlan)
```
Comment (hanfei1991) [inline plan/decorrelate.go]: line 90's innerPlan is apply.children[1], at this place , it will be replaced by agg's child.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**